Sources/TLS/Socket.swift

```diff
@@ -1,174 +1,176 @@
+        )
     }
     
-    /**
-     Accepts a connection to this SSL server from a client.
-     
-     This should only be called if the Context's mode is `.server`
-     */
+    /// Accepts a connection to this SSL server from a client.
+    ///
+    /// This should only be called if the Context's mode is `.server`
     public func accept() throws {
-        let new = try socket.accept()
-        let result = tls_accept_socket(
-            config.context.cContext,
-            &currContext,
-            new.descriptor
-        )
-        currSocket = new
-        
-        guard result == Result.OK else {
-            try new.close()
-            throw TLSError.accept(config.context.error)
-        }
-        
-        // handshake is performed automatically when using tls_read or tls_write, but by doing it here, handshake errors can be properly reported
-        guard tls_handshake(currContext) == Result.OK else {
-            try new.close()
-            throw TLSError.handshake(lastError ?? "Unknown")
+        let client = try socket.accept()
+
+        guard let ssl = SSL_new(context.cContext) else {
+            throw makeError(functionName: "SSL_new")
         }
+        cSSL = ssl
+
+        try assert(
+            SSL_set_fd(ssl, client.descriptor),
+            functionName: "SSL_set_fd"
+        )
+        // keep from deallocating
+        self.client = client
+
+        try assert(
+            SSL_accept(ssl),
+            functionName: "SSL_accept"
+        )
+
+        try assert(
+            SSL_do_handshake(ssl),
+            functionName: "SSL_do_handshake"
+        )
     }
     
-    /**
-     Receives bytes from the secure socket.
-     
-     - parameter max: The maximum amount of bytes to receive.
-     */
-    public func receive(max: Int) throws -> [UInt8]  {
-        guard let context = currContext else {
-            throw TLSError.receive("Context is nil")
-        }
-        
-        let pointer = UnsafeMutablePointer<UInt8>.allocate(capacity: max)
+    /// Receives bytes from the secure socket.
+    ///
+    /// - parameter max: The maximum amount of bytes to receive.
+    public func receive(max: Int) throws -> Bytes  {
+        let pointer = UnsafeMutablePointer<Byte>
+            .allocate(capacity: max)
         defer {
             pointer.deallocate(capacity: max)
         }
+
+        let bytesRead = SSL_read(cSSL, pointer, Int32(max))
         
-        let result = tls_read(context, pointer, max)
-        let bytesRead = Int(result)
-        
-        guard bytesRead >= 0 else {
-            throw TLSError.receive(lastError ?? "Unknown")
+        if bytesRead <= 0 {
+            throw makeError(
+                functionName: "SSL_read",
+                returnCode: bytesRead
+            )
         }
         
-        let buffer = UnsafeBufferPointer<UInt8>.init(start: pointer, count: bytesRead)
+        let buffer = UnsafeBufferPointer<UInt8>.init(start: pointer, count: Int(bytesRead))
```

Comment:
I think `<Byte>(start: ` since we use that one a few lines later, might be worth making `ByteBuffer` type alias, don't worry about this if you want to merge